keps/sig-autoscaling/draft-20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,277 @@
+  (if ResourceRequirements differs from ResourceAllocated),
+* Awaiting - awaiting resources to be freed (e.g. via pre-emption),
+* Failed - resource update could not have been performed in-place
+  but might be possible if some conditions change,
+* Rejected - resource update was rejected by any of the components involved.
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizingPolicy flag,
+available per each resource request (CPU, memory) :
+* InPlace - the default value; allow in-place resize of the Container,
+* RestartContainer - restart the Container to apply new resource values
+  (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart whole Pod to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizingPolicy flag the user can mark Containers or Pods as safe
+(or unsafe) for in-place resources update.
+
+This flag **may** be used by the actors starting the process to decide if
+the process should be started at all (for example VPA might decide to
+evict Pod with RestartPod policy).
+This flag **must** be used by Kubelet to verify the actions needed.
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problems whereas
+changes to available memory are more probable to require restarts.
+
+### Flow Control
+
+The following steps denote a positive flow of an in-place update,
+for a Pod having ResizingPolicy set to InPlace for all its Containers.
+Some alternative flows are given in indented steps,
+unless noted otherwise they abort the flow.
+
+1. The initiating actor updates ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements
+   (e.g. limits are not below requested resources, QoS class does not change).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   1. If any of the controllers rejects the update,
+      the InPlaceResize PodCondition is set to Rejected.
+1. API Server updates the PodSpec object and clears InPlaceResize condition.
+1. Scheduler observes that ResourceRequirements and ResourceAllocated differ.
+   It updates its resource cache to use max(ResourceRequirements, ResourceAllocated).
+   1. If required it pre-empts lower-priority Pods, setting
+      the InPlaceResize PodCondition to Awaiting.
+      Once the lower-priority Pods are evicted, Scheduler clears
+      the InPlaceResize PodCondition and the flow continues.
+1. Kubelet observes that ResourceRequirements and ResourceAllocated differ
+   and the InPlaceResize condition is clear.
+   This is done potentially prior to Scheduler pre-empting lower-priority Pods.
+   1. Kubelet checks that new ResourceRequirements do not fit Node’s
+      allocatable resources and sets the InPlaceResize condition to Failed.
+1. Kubelet applies new resource values to cgroups, updates values
```

Comment:
how does this work for memory?  Reducing the memory cgroup limit does not always stick.  Is this converging until it does stick?

Reply:
Please comment on impacts of emptyDir volumes backed by memory and resizing.  There are challenges where charges are transferred to the pod level cgroup from the container.

Reply:
is there an order the kubelet should follow for applying the resize?  Do we resize container cgroups than pod cgroups or pod cgroups (to desired sum) and then container cgroups?  I think ordering may matter in the case of memory where you would want to update the bounding pod first before it’s nested Container’s to account for things like memory backed volumes.

What happens if I resize two containers in a pod simultaneously and one can resize properly and the other cannot?  Probably a dedicated section for this aspect of the design would be worth calling out.

Reply:
How are extension resources handled?

Reply:
@derekwaynecarr Thanks for drawing attention to these key details.
For ordering, my thoughts are to increase pod-level resources first (cpu/memory) and then the containers-level if we have net increase at pod-level. For net decrease at pod-level, we go the other way, adjusting the pod level value at the very end. I experimented with this running the stress tool in the container, and found that if we lowered pod-level before container-level, kubelet kills container for exceeding hard limits. With multiple containers, we should apply the resource decreases ahead of increases. I'm not sure how best to handle failures at cgroup update besides the brute-force way of restarting the container that fails.

I'll also call out in non-goals that extended resources is not in scope for this KEP because VPA does not support it today. @bskiba mentioned during kubecon that if there is enough interest, VPA can consider adding it. On the surface, the flow for updating extended resource allocations would be similar to what we have here, with scheduler approving and kubelet confirming based on the outcomes of their predicate checks. I'm sure devil is in the details, but this model should work.

I'll follow up with you tomorrow's in sig meeting to better understand the emptyDir and memory reduction issue, where I'm wondering if force_empty can help.

Reply:
@derekwaynecarr I apologize this is taking a little longer with my focus divided on another work-item. After prototyping with the ideas in this merged KEP, if we encounter error in multi-container case where one successfully updates its limits, and another does not, I feel the most reasonable course is to in-place restart the container for which UpdateContainerResources failed. When it gets to this point, both scheduler and kubelet have agreed to the in-place resize capacity (request), and it is the limits that need to be set. The alternative to roll back is not very appealing as the rollback could fail too. What are your thoughts about this approach?

Reply:
@derekwaynecarr Mechanism for handling memory decrease and update ordering sequence have been captured in the KEP. I'll resolve this issue, please reopen if there are still any concerns.